### PR TITLE
Add AVIF support through libavif.

### PR DESCRIPTION
**Pull request by vrabaud**
This is to fix https://github.com/opencv/opencv/issues/19271
Extra: https://github.com/opencv/opencv_extra/pull/1069

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by vrabaud** [comment]
I fixed the EXIF that was not handled and I set the output buffers to 16 bit when reading 10/12 bit. We lost the bit depth information though.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by asmorkalov** [comment]
Build with default Ubuntu 22.04 libavif package fails with compilation error:
```
[ 90%] Building CXX object modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/grfmt_avif.cpp.o
/home/arina/Projects/opencv/modules/imgcodecs/src/grfmt_avif.cpp: In constructor ‘cv::AvifDecoder::AvifDecoder()’:
/home/arina/Projects/opencv/modules/imgcodecs/src/grfmt_avif.cpp:139:27: error: ‘avifDecoder’ {aka ‘struct avifDecoder’} has no member named ‘allowIncremental’
  139 |   if (decoder_) decoder_->allowIncremental = AVIF_TRUE;
      |                           ^~~~~~~~~~~~~~~~
make[3]: *** [modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/build.make:104: modules/imgcodecs/CMakeFiles/opencv_imgcodecs.dir/src/grfmt_avif.cpp.o] Error 1
```

File: modules/imgcodecs/test/test_avif.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    for (int channels : {1, 3, 4}) {
+      for (int bit_depth : {8, 10, 12}) {
+        const Key key{channels, bit_depth};
+        imgs_[key] = modifyImage(img_resized, channels, bit_depth);
+      }
+    }
+  }
+
+  static int kWidth;
+  static int kHeight;
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp:65]: Done

File: modules/imgcodecs/test/test_avif.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+ private:
+  typedef std::tuple<int, int> Key;
+  static std::map<Key, cv::Mat> imgs_;
+};
+std::map<std::tuple<int, int>, cv::Mat>
+    Imgcodecs_Avif_Image_EncodeDecodeSuite::imgs_;
+int Imgcodecs_Avif_Image_EncodeDecodeSuite::kWidth = 50;
+int Imgcodecs_Avif_Image_EncodeDecodeSuite::kHeight = 50;
+
+TEST_P(Imgcodecs_Avif_Image_EncodeDecodeSuite, encode_decode) {
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp:76]: Done

File: modules/imgcodecs/test/test_avif.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    EXPECT_EQ(Imgcodecs_Avif_Image_EncodeDecodeSuite::kWidth, img.cols);
+    EXPECT_EQ(Imgcodecs_Avif_Image_EncodeDecodeSuite::kHeight, img.rows);
+
+    if (imread_mode == IMREAD_UNCHANGED) {
+      ASSERT_EQ(img_original.type(), img.type());
+      // Lossless.
+      if (quality == 100) {
+        EXPECT_EQ(0, cvtest::norm(img, img_original, NORM_INF));
+      }
+    }
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp:124]: Done

File: modules/imgcodecs/test/test_avif.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  ASSERT_EQ(anim_original.size(), anim_read.size());
+  for (size_t i = 0; i < anim_read.size(); ++i) {
+    const cv::Mat img = anim_read[i];
+    ASSERT_FALSE(img.empty());
+    EXPECT_EQ(Imgcodecs_Avif_Animation_EncodeDecodeSuite::kWidth, img.cols);
+    EXPECT_EQ(Imgcodecs_Avif_Animation_EncodeDecodeSuite::kHeight, img.rows);
+
+    if (imread_mode == IMREAD_UNCHANGED) {
+      ASSERT_EQ(anim_original[i].type(), img.type());
+    }
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp:223]: Done

File: modules/imgcodecs/test/test_avif.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+INSTANTIATE_TEST_CASE_P(
+    Imgcodecs_AVIF, Imgcodecs_Avif_Animation_EncodeDecodeSuite,
+    ::testing::Combine(::testing::ValuesIn({8, 10, 12}),
+                       ::testing::ValuesIn({1, 3, 4}),
+                       ::testing::ValuesIn({50}),
+                       ::testing::ValuesIn({IMREAD_UNCHANGED, IMREAD_GRAYSCALE,
+                                            IMREAD_COLOR})));
+
+}  // namespace
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp:235]: That would require me to add an AVIF file. I can probably put a small one but that would be 500 bytes. Is that ok ?

File: CMakeLists.txt [hunk 6]
```diff
@@ -1379,6 +1381,10 @@ if(WITH_WEBP OR HAVE_WEBP)
   status("    WEBP:" WEBP_FOUND THEN "${WEBP_LIBRARY} (ver ${WEBP_VERSION})" ELSE "build (ver ${WEBP_VERSION})")
 endif()
 
+if(WITH_AVIF OR HAVE_AVIF)
+  status("    AVIF:" AVIF_FOUND THEN "${AVIF_LIBRARY} (ver ${AVIF_VERSION})" ELSE "build (ver ${AVIF_VERSION})")
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply CMakeLists.txt:1385]: Done

**Comment by asmorkalov** [comment]
@vrabaud Thanks a lot for the update. Looks like you forgot about opencv_extra patch:
```
[ RUN      ] Imgcodecs_Hdr.regression
imread_('/home/arina/Projects/opencv_extra/testdata/highgui//readwrite/rle.hdr'): can't read header: OpenCV(4.7.0-dev) /home/arina/Projects/opencv/modules/imgcodecs/src/grfmt_avif.cpp:183: error: (-212:Parsing error)  BMFF parsing failed in function 'readHeader'

/home/arina/Projects/opencv/modules/imgcodecs/test/test_grfmt.cpp:373: Failure
Value of: img_rle.empty()
  Actual: true
Expected: false
Could not open /home/arina/Projects/opencv_extra/testdata/highgui//readwrite/rle.hdr
[  FAILED  ] Imgcodecs_Hdr.regression (0 ms)
[ RUN      ] Imgcodecs_Hdr.regression_imencode
imread_('/home/arina/Projects/opencv_extra/testdata/highgui//readwrite/rle.hdr'): can't read header: OpenCV(4.7.0-dev) /home/arina/Projects/opencv/modules/imgcodecs/src/grfmt_avif.cpp:183: error: (-212:Parsing error)  BMFF parsing failed in function 'readHeader'

/home/arina/Projects/opencv/modules/imgcodecs/test/test_grfmt.cpp:396: Failure
Value of: img_ref.empty()
  Actual: true
Expected: false
Could not open /home/arina/Projects/opencv_extra/testdata/highgui//readwrite/rle.hdr
[  FAILED  ] Imgcodecs_Hdr.regression_imencode (0 ms)
```

**Comment by vrabaud** [comment]
All imgcodecs tests now pass.

File: cmake/OpenCVFindLibsGrfmt.cmake [hunk 7]
```diff
@@ -37,6 +37,16 @@ if(NOT ZLIB_FOUND)
...
 
+# --- libavif (optional) ---
+
+if(WITH_AVIF)
+  ocv_clear_internal_cache_vars(AVIF_LIBRARY AVIF_INCLUDE_DIR)
+  include(cmake/OpenCVFindAVIF.cmake)
+  if(AVIF_FOUND)
+    set(HAVE_AVIF 1)
+  endif()
+endif()
```
[the remark this answers is not in the document]
Comment (asmorkalov) [reply cmake/OpenCVFindLibsGrfmt.cmake:48]: For WEBP it's most probably mistake.

**Comment by asmorkalov** [comment]
All tests passed on my side with Ubuntu 22.04. Thanks a lot! I'll try to look into code in details for the next several days and provide feedback. I also added Ubuntu 22.04 config with libavif on board to CI in https://github.com/opencv/ci-gha-workflow/pull/101/ and https://github.com/opencv-infrastructure/opencv-gha-dockerfile/pull/18

File: cmake/OpenCVFindAVIF.cmake [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    GET_TARGET_PROPERTY(AVIF_INCLUDE_DIR1 avif INCLUDE_DIRECTORIES)
+    GET_TARGET_PROPERTY(AVIF_INCLUDE_DIR2 avif INTERFACE_INCLUDE_DIRECTORIES)
+    set(AVIF_INCLUDE_DIR)
+    if(AVIF_INCLUDE_DIR1)
+        LIST(APPEND AVIF_INCLUDE_DIR ${AVIF_INCLUDE_DIR1})
+    endif()
+    if(AVIF_INCLUDE_DIR2)
+        LIST(APPEND AVIF_INCLUDE_DIR ${AVIF_INCLUDE_DIR2})
+    endif()
+else()
```
Comment (asmorkalov) [inline cmake/OpenCVFindAVIF.cmake:32]: The first branch does not work. Nobody defines avif target. It makes sense to use `AVIF_FOUND`, if  you uncomment find_package.

File: CMakeLists.txt [hunk 9]
```diff
@@ -1379,6 +1381,14 @@ if(WITH_WEBP OR HAVE_WEBP)
   status("    WEBP:" WEBP_FOUND THEN "${WEBP_LIBRARY} (ver ${WEBP_VERSION})" ELSE "build (ver ${WEBP_VERSION})")
 endif()
 
+if(WITH_AVIF OR HAVE_AVIF)
+  if(AVIF_VERSION)
+    status("    AVIF:" AVIF_FOUND THEN "${AVIF_LIBRARY} (ver ${AVIF_VERSION})" ELSE "NO")
+  else()
+    status("    AVIF:" AVIF_FOUND THEN "${AVIF_LIBRARY}" ELSE "NO")
+  endif()
+endif()
```
Comment (asmorkalov) [inline CMakeLists.txt:1390]: In my case `AVIF_LIBRARY` resolves to target name "avif". It's not informative. I propose to check `if TARGET {AVIF_LIBRARY}` and get target file name for output.

File: modules/imgcodecs/test/test_avif.cpp [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+    EXPECT_NO_FATAL_FAILURE(
+        cv::imwrite(output, img_original, encoding_params_));
+    return;
+  }
+  EXPECT_NO_THROW(cv::imwrite(output, img_original, encoding_params_));
+
+  // Read from file.
+  const cv::Mat img = cv::imread(output, imread_mode_);
+
+  ValidateRead(img_original, img);
```
Comment (asmorkalov) [inline modules/imgcodecs/test/test_avif.cpp:143]: temporary file is not removed in the test.

File: cmake/OpenCVFindAVIF.cmake (hunk 8 above)
Comment (vrabaud) [reply cmake/OpenCVFindAVIF.cmake:32]: My bad, find_package was commented out for debug ....

File: CMakeLists.txt (hunk 9 above)
Comment (vrabaud) [reply CMakeLists.txt:1390]: Fixed

File: modules/imgcodecs/src/grfmt_avif.cpp [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+}
+
+bool AvifEncoder::write(const Mat &img, const std::vector<int> &params) {
+  std::vector<Mat> img_vec(1, img);
+  return writeToOutput(img_vec, params);
+}
+
+bool AvifEncoder::writemulti(const std::vector<Mat> &img_vec,
+                             const std::vector<int> &params) {
+  encoder_->timescale = 10;
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply modules/imgcodecs/src/grfmt_avif.cpp]: Removed: this is related to the frequency of the animation. We have no parameter for that in OpenCV. Ideally, there would be a frame duration.

File: modules/imgcodecs/test/test_avif.cpp (hunk 10 above)
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp:142]: Fixed

File: modules/imgcodecs/test/test_avif.cpp [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+  image->transferCharacteristics = AVIF_TRANSFER_CHARACTERISTICS_UNSPECIFIED;
+  image->matrixCoefficients = AVIF_MATRIX_COEFFICIENTS_IDENTITY;
+  image->yuvRange = AVIF_RANGE_FULL;
+  image->yuvPlanes[0] = img_in.data;
+  image->yuvRowBytes[0] = img_in.step[0];
+  image->imageOwnsYUVPlanes = AVIF_FALSE;
+
+  constexpr uint8_t kExif[] = {73, 73, 42, 0, 8, 0, 0, 0, 1, 0, 18, 1, 3,
+                               0,  1,  0,  0, 0, 5, 0, 0, 0, 0, 0,  0, 0};
+  avifImageSetMetadataExif(image, kExif, 26);
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp]: Fixed

[the remark this answers is not in the document]
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp]: The files were added in https://github.com/opencv/opencv_extra/pull/1068

**Comment by vrabaud** [comment]
(the CI error seems unrelated: DNN with CUDA error)

File: modules/imgcodecs/test/test_avif.cpp [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+  static const int kHeight;
+
+ private:
+  typedef std::tuple<int, int> Key;
+  static std::map<Key, cv::Mat> imgs_;
+};
+std::map<std::tuple<int, int>, cv::Mat>
+    Imgcodecs_Avif_Image_RoundTripSuite::imgs_;
+const int Imgcodecs_Avif_Image_RoundTripSuite::kWidth = 51;
+const int Imgcodecs_Avif_Image_RoundTripSuite::kHeight = 51;
```
Comment (asmorkalov) [inline modules/imgcodecs/test/test_avif.cpp:120]: It makes sense to make them different to ensure dimentions swap issue.

**Review by asmorkalov — APPROVED** [review]
:+1:

File: modules/imgcodecs/test/test_avif.cpp (hunk 13 above)
Comment (vrabaud) [reply modules/imgcodecs/test/test_avif.cpp:120]: Done

**Comment by asmorkalov** [comment]
@vrabaud We discussed the solution on core team meeting and decided to disable libavif by default for now and treat it's support as experimental till 4.9.0. Could you disable the option by default? I'll add CI configuration a bit later after the release.

**Comment by vrabaud** [comment]
@asmorkalov: sure, done. WITH_AVIF is set to OFF by default.

**Comment by umbertov** [comment]
@asmorkalov @vrabaud  Hey, I was reading cmake files and noticed  that AVIF support relies on headers & libraries either being  present on the system, or built separately and pointed at by the user, while many other 3rd party components like jpeg, png, webp and more have an option (like `BUILD_JPEG` for libjpeg, for instance) to be built together with opencv and being statically linked into the `libopencv_*.so` libraries. I assume that they're being statically linked since `ldd` shows no references to e.g. libjpeg.

Is a `BUILD_AVIF` option not implemented just because of time & resources reasons, or because of weird licensing issues? Because if it is just due to the former, i might try to work on implementing it.
Thanks in advance for any answer!

**Comment by asmorkalov** [comment]
Major issue is mostly engineering. AVIF has a lot of own dependencies that we do not want to bring into OpenCV. Licenses for AVIF dependence sub-tree is open issue too.

**Comment by vrabaud** [comment]
Licenses are fine: libavif is BSD 2 clause.
Now, it is a thin wrapper around any encoding and/or decoding library.
If we pull in libaom, it is 400 files, dav1d (only for decoding) 120. That is a lot to add to the third_party directory: that was the only rational not to include it.

**Pull request merged**